### PR TITLE
fix #158589

### LINKED ISSUE #158589: tasks aren't reconnected to
on window reload

```
 INFO getting task system before reconnection
log.ts:301  INFO attempting task reconnection, jsonTasksSupported: false, reconnection pending true
log.ts:301  INFO attempting task reconnection, jsonTasksSupported: true, reconnection pending false
```

### PR DESCRIPTION
**Pull request by meganrogge**
fix #158589

### CODE REVIEW

File: src/vs/workbench/contrib/tasks/browser/terminalTaskSystem.ts [hunk 1]
```diff
@@ -1297,6 +1297,7 @@ export class TerminalTaskSystem extends Disposable implements ITaskSystem {
 		for (let i = 0; i < this._reconnectedTerminals.length; i++) {
 			const terminal = this._reconnectedTerminals[i];
 			const taskForTerminal = terminal.shellLaunchConfig.attachPersistentProcess?.reconnectionProperties?.data as IReconnectionTaskData;
+			console.log('reconnected terminal', taskForTerminal, taskForTerminal.lastTask, task.getMapKey());
```
Comment (meganrogge) [inline src/vs/workbench/contrib/tasks/browser/terminalTaskSystem.ts:1300]:
```suggestion
```

**Review by alexr00 — DISMISSED** [review]

**Review by Tyriar — DISMISSED** [review]

File: src/vs/workbench/contrib/tasks/browser/abstractTaskService.ts [hunk 2]
```diff
@@ -390,24 +389,27 @@ export abstract class AbstractTaskService extends Disposable implements ITaskSer
...
-			}
-			if (result?.exitCode === terminalsNotReconnectedExitCode) {
-				this._logService.trace('Terminals were not reconnected');
-				return false;
+				result = this.run(task, undefined, TaskRunSource.Reconnect);
 			}
+			promises.push(result);
+		}
+		const promiseResult = await Promise.all(promises);
+		if (promiseResult.find(p => p?.exitCode === terminalsNotReconnectedExitCode)) {
```
Comment (meganrogge) [inline src/vs/workbench/contrib/tasks/browser/abstractTaskService.ts:411]: note that this case shouldn't happen anymore, so probably safe to remove during debt week

**Comment by meganrogge** [comment]
works on first reload, but on second - only 1 of the two task terminals gets reconnected to.

interestingly, when I delete the one that does get reconnected, the other one gets reconnected. I believe this has to do with waiting for the promises

**Review by Tyriar — DISMISSED** [review]

**Comment by meganrogge** [comment]
this works beautifully for the first window reload, but not the second 😢 it ends up running the tasks again instead of reconnecting - probably because the terminals haven't reconnected when the first attempt occurs

**Comment by meganrogge** [comment]
which means I think I do need the exit code stuff that I removed

**Comment by meganrogge** [comment]
nope doesn't work in the build 🤷🏼‍♀️  maybe we can look at this timing thing together on Monday @Tyriar ? 

we really need to create the task system immediately. then wait for supported executions to be available and wait for terminals before trying to reconnect and only do so once

File: src/vs/workbench/contrib/tasks/browser/terminalTaskSystem.ts [hunk 3]
```diff
@@ -1339,22 +1330,20 @@ export class TerminalTaskSystem extends Disposable implements ITaskSystem {
 
 	private _reconnectToTerminals(): void {
 		if (this._hasReconnected) {
-			this._logService.trace(`Already reconnected, to ${this._reconnectedTerminals?.length} terminals so returning`);
+			this._logService.info(`Already reconnected, to ${this._reconnectedTerminals?.length} terminals so returning`);
```
Comment (Tyriar) [inline src/vs/workbench/contrib/tasks/browser/terminalTaskSystem.ts:1333]: We should avoid info as it will show up all the time

**Review by Tyriar — APPROVED** [review]

File: src/vs/workbench/contrib/tasks/browser/terminalTaskSystem.ts (hunk 3 above)
Comment (meganrogge) [reply src/vs/workbench/contrib/tasks/browser/terminalTaskSystem.ts:1333]: Yes I'll fix that Monday

**Comment by meganrogge** [comment]
@meganrogge Create issue for changing log level to trace

**Pull request merged**